fishPlugins: create scope and add some plugins

pacien opened the pull request:
###### Motivation for this change

This bootstraps the fishPlugins scope and build tools for packaging plugins for the fish shell.

Two plugin packages are added:
* [fishtape](https://github.com/jorgebucaran/fishtape): a TAP-based test runner for Fish, often used to test fish plugins,
* [pure](https://github.com/rafaelrinaldi/pure): a pretty, minimal and fast Fish prompt, ported from zsh


###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).


###### Poking people

CC'ing the most recent committers on the fish module and package as they might be interested in this PR:
* @cole-h
* @gebner
* @adisbladis
* @bouk
* @xaverdh
* @rnhmjoj
* @hyperfekt

pacien committed:
fishPlugins: bootstrap fish plugin scope

pacien committed:
fishPlugins.fishtape: init at 2.1.3

pacien committed:
fishPlugins.pure: init at 3.4.2

rnhmjoj commented on pkgs/top-level/all-packages.nix:
```diff
@@ -8744,6 +8744,8 @@ in
 
+  fishPlugins = recurseIntoAttrs (callPackage ../misc/fish-plugins { });
+
   fish-foreign-env = callPackage ../shells/fish/fish-foreign-env { };
```
Can you move `foreign-env` into fish-plugins? It has received a special treatment so far, because it's needed to source the NixOS envronment.

bouk replied:
Kind of tangental, I created https://github.com/bouk/babelfish to avoid having fish-foreign-env in my shell, because it's super slow in my experience. Spawning a prompt takes more than a second. Might be worth integrating in the future.

And I agree we should move ffe into the fishPlugins scope

pacien replied:
Package moved and module updated :+1:

pacien committed:
fish-foreign-env: move to fishPlugins.foreign-env

And relocate the installed fish functions to the `vendor_functions.d` so
that they're automatically loaded.

rnhmjoj commented on pkgs/top-level/all-packages.nix:
```diff
@@ -8744,7 +8744,11 @@ in
-  fish-foreign-env = callPackage ../shells/fish/fish-foreign-env { };
+  fishPlugins = recurseIntoAttrs (callPackage ../misc/fish-plugins { });
+
+  fish-foreign-env = throw ''
```
You should define an alias in `pkgs/top-level/aliases.nix`, otherwise this will break existing setups.

pacien replied:
Amended :+1:

rnhmjoj commented on pkgs/top-level/aliases.nix:
```diff
@@ -156,6 +156,7 @@ mapAliases ({
   firefoxWrapper = firefox;           # 2015-09
 
   firestr = throw "firestr has been removed."; # added 2019-12-08
+  fish-foreign-env = fishPlugins.fish-foreign-env;
```
Also add a date, please.

pacien replied:
Added and amended.

rnhmjoj commented:
It would be good to add documentation on how to add/use these plugins.
I can't figure how to use them: either with adding `fishPlugins.something` to `environment.systemPackages` or a nix-shell fish doesn't seems to find them, even with `programs.fish.vendor.config.enable = true`.

pacien commented:
That's strange. This should be enough for the plugin to be detected by fish.

I'm using fish enabled through home-manager's `programs.fish.enable` and adding the packages to `home.packages` is enough for them to be detected.

rnhmjoj commented:
> I'm using fish enabled through home-manager's programs.fish.enable and adding the packages to home.packages is enough for them to be detected.

Uhm, I don't use home-manager, so I tried:
```
nix-shell -I nixpkgs=/path/to/nixpkgs --pure -p fish fishPlugins.fish-foreign-env -- run fish
```
but `fenv` is not available. Similarly, when I try a NixOS VM with:
```
nix-build -E '(import ./nixos { 
  configuration = { pkgs,... }: { 
    users.users.root.password = "root";
    programs.fish.enable = true;
    environment.systemPackages = with pkgs; [ fishPlugins.pure fishPlugins.fish-foreign-env ];
  }; }).vm'
```

pacien commented:
It seems like the `fish-foreign-env` package was installing the scripts outside of the standard `vendor_functions.d` directory.
I modified the package and the module accordingly and everything is correctly loaded now.

This is however a breaking change. Should I remove the package alias?

rnhmjoj commented:
> This is however a breaking change. Should I remove the package alias?

Uhm, it's likely it was never used outside of the NixOS module.
I'd say to add a note in the release notes but keep the alias.

pacien commented:
Note added to the release notes and amended.

rnhmjoj commented:
Thank you. I tested the NixOS module and built the manual: looks all good.

rnhmjoj commented on pkgs/misc/fish-plugins/build-fish-plugin.nix:
```diff
@@ -0,0 +1,77 @@
+    then v
+    else "${v}/${rtpPath}/vendor_functions.d";
+
+  fishWithFunctionPath = plugins: let
```
What about using this function to define a `fish.withPlugins`? It would be useful for quickly testing plugins and running them inside a nix-shell.

pacien replied:
Good idea, this would be useful.

I think it can be done by creating some `wrapFish` package which would accept lists of plugins, custom functions, completion scripts, initialisation scripts and environment variables as customisation parameters.

This is a bit involved though and I'd prefer to keep that for a separate pull request.

rnhmjoj replied:
> This is a bit involved though and I'd prefer to keep that for a separate pull request.

Yeah, I agree.

rnhmjoj reviewed (approved):
I don't have any other suggestions to make, the changes look good to me.
I'd wait for more fish users/maintainers to comment before merging, though.

cole-h commented on pkgs/top-level/all-packages.nix:
```diff
@@ -8744,7 +8744,7 @@ in
   fish = callPackage ../shells/fish { };
 
-  fish-foreign-env = callPackage ../shells/fish/fish-foreign-env { };
+  fishPlugins = recurseIntoAttrs (callPackage ../misc/fish-plugins { });
```
I think this should go into a new `shells/fish/plugins` directory rather than `misc/fish-plugins`.

pacien replied:
Indeed. I moved the directory as suggested and amended.

cole-h reviewed (changes requested):
One small suggestion below, otherwise LGTM. Thanks for this, @pacien! And thanks @rnhmjoj for all your feedback / review.

cole-h reviewed (approved):
Diff LGTM. Thanks!

pacien commented:
(Rebased to fix merrge conflict.)

abathur commented:
@Pacien I'm not a fish user so I don't have much insight, but it sounds like this breaking change is sneaking up on some nix-darwin users (https://github.com/LnL7/nix-darwin/issues/269). I would guess it's doing the same for HM users?

rnhmjoj commented:
Yeah, nix-darwin and probably home-manager need to change the foreign-env path as in this diff:
https://github.com/pacien/nixpkgs/commit/d94921db12d9c9e2d3e9c2e6e25e50f03f789ecb#diff-e643907e997ff11fbccddd7dacf176032a320f26f8313eb4b18f435597dca2b0

pacien commented:
Should we remove the alias `fish-foreign-env` -> `fishPlugins.foreign-env` so that the evaluation fails instead of having runtime breakage? Those packages aren't compatible due to the directory layout changes, so I'm not sure that it made sense to have that alias in the first place.

cole-h commented:
Yeah, I agree with that.
